Implement rustc side of report-future-incompat

Aaron1011 opened the pull request:
cc https://github.com/rust-lang/rust/issues/71249

This is an alternative to @pnkfelix's initial implementation in https://github.com/pnkfelix/rust/commits/prototype-rustc-side-of-report-future-incompat (mainly because I started working before seeing that branch :smile: ).

My approach outputs the entire original `Diagnostic`, in a way that is compatible with incremental compilation. This is not yet integrated with compiletest, but can be used manually by passing `-Z emit-future-incompat-report` to `rustc`.

Several changes are made to support this feature:
* The `librustc_session/lint` module is moved to a new crate `librustc_lint_defs` (name bikesheddable). This allows accessing lint definitions from `librustc_errors`.
* The `Lint` struct is extended with an `Option<FutureBreakage>`. When present, it indicates that we should display a lint in the future-compat report. `FutureBreakage` contains additional information that we may want to display in the report (currently, a `date` field indicating when the crate will stop compiling).
* A new variant `rustc_error::Level::Allow` is added. This is used when constructing a diagnostic for a future-breakage lint that is marked as allowed (via `#[allow]` or `--cap-lints`). This allows us to capture any future-breakage diagnostics in one place, while still discarding them before they are passed to the `Emitter`.
* `DiagnosticId::Lint` is extended with a `has_future_breakage` field, indicating whether or not the `Lint` has future breakage information (and should therefore show up in the report).
* `Session` is given access to the `LintStore` via a new `SessionLintStore` trait (since `librustc_session` cannot directly reference `LintStore` without a cyclic dependency). We use this to turn a string `DiagnosticId::Lint` back into a `Lint`, to retrieve the `FutureBreakage` data.

Currently, `FutureBreakage.date` is always set to `None`. However, this could potentially be interpreted by Cargo in the future.

I've enabled the future-breakage report for the `ARRAY_INTO_ITER` lint, which can be used to test out this PR. The intent is to use the field to allow Cargo to determine the date of future breakage (as described in [RFC 2834](https://github.com/rust-lang/rfcs/blob/master/text/2834-cargo-report-future-incompat.md)) without needing to parse the diagnostic itself.

cc @pnkfelix

rust-highfive commented:
r? @ecstatic-morse

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #73851) made this pull request unmergeable. Please resolve the merge conflicts.

ecstatic-morse commented:
Not familiar with this part of the code.

r? @pnkfelix maybe?

bors commented:
:umbrella: The latest upstream changes (presumably #75516) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #75944) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74862) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #76231) made this pull request unmergeable. Please resolve the merge conflicts.

Aaron1011 commented:
@pnkfelix: This is now ready for review

bors commented:
:umbrella: The latest upstream changes (presumably #76549) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
@pnkfelix: Are there any changes that you'd like me to make?

bors commented:
:umbrella: The latest upstream changes (presumably #70743) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
@pnkfelix: Do you think you'll be able to review this sometime soon? There are several issues that would benefit from having this available.

bors commented:
:umbrella: The latest upstream changes (presumably #77557) made this pull request unmergeable. Please resolve the merge conflicts.

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

pnkfelix commented:
@Aaron1011 my personal preference would be for this to be broken into more commits; e.g. the initial factoring of `rustc_lint_defs` into its own crate, without adding any new functionality (such as variants to enums it defines), could be a separate commit.

My main motivation for splitting it up into commits like that is that its easier to distinguish the pure refactorings from the commits that actually add functionality and/or change behavior, which in turn allows my brain to focus on different kinds of bugs to look for.

But I will still take a look at it as-is, since its not an enormous revision and so I can hopefully still provide useful feedback on the approach as written.

pnkfelix commented on compiler/rustc_errors/src/json.rs:
```diff
@@ -131,6 +132,30 @@ impl Emitter for JsonEmitter {
+            writeln!(&mut self.dst, "{}", as_pretty_json(&data))
+        } else {
+            writeln!(&mut self.dst, "{}", as_json(&data))
+        }
```
should we consider a mode (not the default; presumably under a `-Z` flag) that emits the `data` in a human readable format?

pnkfelix commented:
Okay, I think the approach given here makes sense to me.

But it needs unit tests in `tests/ui`; each lint that we revise to use this framework will want to add its own tests of the functionality. And that in turn, I think, will motivate the human readable output format that I alluded to above (this human readable format would be enabled via a `-Z` flag, and then the `ui` test will enable that flag.

Aaron1011 commented:
@pnkfelix: I don't think having human-readable output emitted by rustc would be very useful. The actual messages displayed to the user will be emitted from Cargo by parsing the json output, which could cause any testing-only output to get out of sync.

I think it would be better to have compiletest perform some minimal parsing of the future-compat report, and add an new `~FUTURE-COMPAT` comment that can be used in test files.

Aaron1011 commented:
@pnkfelix: I ended up going with a combination of our ideas. rustc only ever emits a JSON future-compat-report, but compiletest renders it a simple human-readable format.

Other changes:
* I renamed the flag from `-Z emit-future-compat-report` to `-Z emit-future-incompat-report` for consistency wit the RFC.
* Compiletest passes `-Z emit-future-compat-report` to every UI test, and skips rendering if the JSON report message is empty.
* I added a workaround for invoking `rustfix` in compiletest - we now filter out all non-`Diagnostic` json lines, since compiletest doesn't recognize the future-incompat-report message.

The test `src/test/ui/iterators/into-iter-on-arrays-lint.rs` shows the results of these changes.

bors commented:
:umbrella: The latest upstream changes (presumably #78324) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

Aaron1011 commented:
@pnkfelix: Are there any changes that you'd like me to make?

bors commented:
:umbrella: The latest upstream changes (presumably #78421) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

*Note that reviewers usually do not review pull requests until merge conflicts are resolved!* Once you resolve the conflicts, you should change the labels applied by bors to indicate that your PR is ready for review. Post this as a comment to change the labels:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

pnkfelix commented:
> @pnkfelix: Are there any changes that you'd like me to make?

This looks good!

The only thing that gives me pause is deciding whether we should be enabling the future incompatibility report generation by default in the UI tests.

But I don't know *why* I'm so concerned about that; its not like its causing a swath of changes to the existing tests.

pnkfelix commented:
r=me after resolving the latest set of merge conflicts.

Aaron1011 commented:
@bors r=pnkfelix

bors commented:
:pushpin: Commit 8cad828bf70e834d784600166c73e28096962543 has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 8cad828bf70e834d784600166c73e28096962543 -->
<!-- homu: {"type":"Approved","sha":"8cad828bf70e834d784600166c73e28096962543","approver":"pnkfelix"} -->

bors commented:
:hourglass: Testing commit 8cad828bf70e834d784600166c73e28096962543 with merge fddef549a61710ee260841a08944b47689e8a2df...
<!-- homu: {"type":"BuildStarted","head_sha":"8cad828bf70e834d784600166c73e28096962543","merge_sha":"fddef549a61710ee260841a08944b47689e8a2df"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1330620279)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1330620279","builder_name":"checks-actions"} -->

Aaron1011 commented:
@bors r=pnkfelix

bors commented:
:pushpin: Commit 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e has been approved by `pnkfelix`

<!-- @bors r=pnkfelix 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e -->
<!-- homu: {"type":"Approved","sha":"6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e","approver":"pnkfelix"} -->

bors commented:
:hourglass: Testing commit 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e with merge 657091462a5bb2c6f196c829e6e526ab835ee153...
<!-- homu: {"type":"BuildStarted","head_sha":"6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e","merge_sha":"657091462a5bb2c6f196c829e6e526ab835ee153"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1335206477)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1335206477","builder_name":"checks-actions"} -->

Aaron1011 commented:
@bors retry

bors commented:
:hourglass: Testing commit 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e with merge 962969d3ff75becab9be769a4f271dad63e0155f...
<!-- homu: {"type":"BuildStarted","head_sha":"6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e","merge_sha":"962969d3ff75becab9be769a4f271dad63e0155f"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

Aaron1011 commented:
@bors retry

bors commented:
:hourglass: Testing commit 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e with merge 7bcd889ebca66a5d5e6224220bc58bca2a240ddf...
<!-- homu: {"type":"BuildStarted","head_sha":"6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e","merge_sha":"7bcd889ebca66a5d5e6224220bc58bca2a240ddf"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1337632420)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1337632420","builder_name":"checks-actions"} -->

Aaron1011 commented:
@bors retry

bors commented:
:hourglass: Testing commit 6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e with merge b2025326088b54fb3f083bebeba14e0a15bf00d3...
<!-- homu: {"type":"BuildStarted","head_sha":"6db00a213a6c8815aed7ae2d4e71d8a1b5b6338e","merge_sha":"b2025326088b54fb3f083bebeba14e0a15bf00d3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1339132986)
Approved by: pnkfelix
Pushing b2025326088b54fb3f083bebeba14e0a15bf00d3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"pnkfelix","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1339132986"},"merge_sha":"b2025326088b54fb3f083bebeba14e0a15bf00d3"} -->
